### ISSUE TITLE
Confusion on to_categorical function for class Train. 

### ISSUE THREAD

**Issue by umesh-timalsina** [KyleAMoore/Redshift-CNN#38]
The `to_categorical` function in class `Train` is supposed to divide the labels distribution into a uniform a categorical bins but its only generating labels between 0 and 3. What is the correct way to use it?

```py
>>> from train import Train
>>> import numpy as np
>>> trainer = Train()
>>> labels = np.random.uniform(low=0.0, high=0.39999999, size=(4000, 1))
>>> cat_labels = trainer.to_categorical(labels)
>>> num_cats_gt_4 = (cat_labels > 4).sum()
>>> num_cats_gt_4
0
>>> num_cats_lt_4 = (cat_labels < 4).sum()
>>> num_cats_lt_4
4000
```

**Comment by umesh-timalsina** [comment]
I guess the other way of framing the question is why is default value for the `max_rs_val` for the class 3.5?

**Comment by KyleAMoore** [comment]
The default value was arbitrarily chosen because it is the closest simple number to the actual highest redshift value of our initial dataset. For other datasets, it may be more appropriate to make it higher or lower, with the presumption that it is best to keep the possible range as close to the actual range as possible to allow the smallest granularity possible with the given number of bins. 

Unfortunately, I currently don't know of a better way to estimate a better default value. I suspect, though have not tested, that setting it to the currently known maximum value ([11.09](https://en.wikipedia.org/wiki/List_of_the_most_distant_astronomical_objects)) could degrade performance by reducing the granularity of the bins. This seems especially important when considering that the majority of the samples that we have already used for training are very low (<1.0). I have considered the possibility of setting the max to slightly above the highest value in the training set, but I am unsure if this could cause problems for galaxies with higher redshift scores if the training set happened to be more skewed towards closer galaxies than the true distribution.

**Comment by umesh-timalsina** [comment]
> The default value was arbitrarily chosen because it is the closest simple number to the actual highest redshift value of our initial dataset. For other datasets, it may be more appropriate to make it higher or lower, with the presumption that it is best to keep the possible range as close to the actual range as possible to allow the smallest granularity possible with the given number of bins.

Isn't it the case that the redshift for the initial dataset is in the range [0, 0.4)? Categorizing them into bins of range [0, 3.5) would mean the distribution would not be representative of the dataset itself.

 > Unfortunately, I currently don't know of a better way to estimate a better default value. I suspect, though have not tested, that setting it to the currently known maximum value (11.09) could degrade performance by reducing the granularity of the bins. This seems especially important when considering that the majority of the samples that we have already used for training are very low (<1.0). I have considered the possibility of setting the max to slightly above the highest value in the training set, but I am unsure if this could cause problems for galaxies with higher redshift scores if the training set happened to be more skewed towards closer galaxies than the true distribution.

Quoting from the redshift paper "From the SDSS database, we retrieved 516,525 sources classified as galaxy, with dereddened petrosian magnitudes `r ≤ 17.8` and spectroscopic redshifts `z ≤ 0.4`. We should use this as the default value. The number of bins should be determined based on the training size.

**Comment by KyleAMoore** [comment]
The initial dataset used by the original authors was using the range `[0,0.4)`, but the smaller starting set I was provided initially and used for testing was in the range `[0,0.344)`, thus the current use of 3.5 as the default max value. I admittedly missed that point in the paper and agree that future versions should use 0.4 for the default. I will fix this in all relevant locations.

**Comment by KyleAMoore** [comment]
As an additional note, I just realized that I was using 3.5 instead of the originally intended 0.35. This appears to have been a slight mistake that I made early on that persisted through numerous versions. This might explain some of the earlier underperformance compared to the original paper.
